File: 02/02.py
from pathlib import Path
from typing import List

POINTS = {"ROCK": 1, "PAPER": 2, "SCISSORS": 3, "X": 1, "Y": 2, "Z": 3}


def play_round(opponent: str, your: str) -> int:
    if (opponent == "A" and your == "X") or (opponent == "B" and your == "Y") or (opponent == "C" and your == "Z"):
        return 3 + POINTS[your]
    if (opponent == "A" and your == "Z") or (opponent == "B" and your == "X") or (opponent == "C" and your == "Y"):
        return POINTS[your]
    if (opponent == "A" and your == "Y") or (opponent == "B" and your == "Z") or (opponent == "C" and your == "X"):
        return 6 + POINTS[your]
    raise ValueError(f"Combination not found {opponent} {your}")


def play_round2(opponent: str, outcome: str) -> int:
    if outcome == "X":  # DEFEAT
        if opponent == "A":
            return POINTS["SCISSORS"]
        if opponent == "B":  # PAPER
            return POINTS["ROCK"]
        return POINTS["PAPER"]  # PAPER
    if outcome == "Y":  # DRAW
        if opponent == "A":  # ROCK
            return 3 + POINTS["ROCK"]
        if opponent == "B":  # PAPER
            return 3 + POINTS["PAPER"]
        return 3 + POINTS["SCISSORS"]
    # WIN
    if opponent == "A":  # ROCK
        return 6 + POINTS["PAPER"]
    if opponent == "B":  # PAPER
        return 6 + POINTS["SCISSORS"]
    return 6 + POINTS["ROCK"]


def first_question(input_data: str) -> int:
    return sum(map(lambda x: play_round(*x.split()), input_data.splitlines()))


def second_question(input_data: str) -> int:
    return sum(map(lambda x: play_round2(*x.split()), input_data.splitlines()))


if __name__ == "__main__":

    input_data = (Path(__file__).parent / "INPUT").read_text("UTF8")
    print(first_question(input_data))
    print(second_question(input_data))

File: 02/02.rs
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

use phf::phf_map;

static POINTS: phf::Map<&'static str, i32> = phf_map!(
    "ROCK" => 1,
    "PAPER" => 2,
    "SCISSORS" => 3,
    "X" => 1,
    "Y" => 2,
    "Z" => 3,
);

fn main() {
    let mut lines = read_lines("./INPUT").expect("error reading file");
    if let Ok(answer1) = first_question(lines) {
        println!("{}", answer1);
    };
    lines = read_lines("./INPUT").expect("error reading file");
    if let Ok(answer2) = second_question(lines) {
        println!("{}", answer2);
    };
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn first_question(lines: io::Lines<io::BufReader<File>>) -> io::Result<i32> {
    let mut points: i32 = 0;
    for line in lines {
        let l = line?;
        let moves: Vec<&str> = l.split_whitespace().collect();
        let opponent = moves[0];
        let your = moves[1];
        let point = POINTS.get(moves[1]).unwrap();
        if (opponent == "A" && your == "X")
            || (opponent == "B" && your == "Y")
            || (opponent == "C" && your == "Z")
        {
            points += 3 + point;
            continue;
        }
        if (opponent == "A" && your == "Z")
            || (opponent == "B" && your == "X")
            || (opponent == "C" && your == "Y")
        {
            points += point;
            continue;
        }
        points += 6 + point;
    }
    return Ok(points);
}

fn second_question(lines: io::Lines<io::BufReader<File>>) -> io::Result<i32> {
    let mut points: i32 = 0;
    for line in lines {
        let l = line?;
        let moves: Vec<&str> = l.split_whitespace().collect();
        let opponent = moves[0];
        let outcome = moves[1];
        if outcome == "X" {
            // DEFEAT
            if opponent == "A" {
                //ROCK
                points += POINTS["SCISSORS"];
                continue;
            }
            if opponent == "B" {
                // PAPER
                points += POINTS["ROCK"];
                continue;
            }
            points += POINTS["PAPER"];
            continue;
        }
        if outcome == "Y" {
            //DRAW
            if opponent == "A" {
                points += 3 + POINTS["ROCK"];
                continue;
            }
            if opponent == "B" {
                // PAPER
                points += 3 + POINTS["PAPER"];
                continue;
            }
            points += 3 + POINTS["SCISSORS"];
            continue;
        }
        // WIN
        if opponent == "A" {
            // ROCK
            points += 6 + POINTS["PAPER"];
            continue;
        }
        if opponent == "B" {
            //PAPER
            points += 6 + POINTS["SCISSORS"];
            continue;
        }
        points += 6 + POINTS["ROCK"];
    }
    return Ok(points);
}
